multidict/_multidict_c.c

```diff
@@ -0,0 +1,854 @@
+
+#define MultiDict_Check(o) (Py_TYPE(o) == &multidict_type)
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *impl;
+} _MultiDictObject;
+
+/******************** Base Methods ********************/
+
+/* Forward declaration */
+static PyObject *multidict_items(_MultiDictObject *self);
+
+static INLINE PyObject *
+_multidict_getone(_MultiDictObject *self, PyObject *key, PyObject *_default)
+{
+    PyObject *val = pair_list_get_one(self->impl, key);
+
+    if (val == NULL &&
+        PyErr_ExceptionMatches(PyExc_KeyError) &&
+        _default != NULL)
+    {
+        PyErr_Clear();
+        return _default;
+    }
+
+    return val;
+}
+
+static PyObject *
+_multidict_eq(_MultiDictObject *self, _MultiDictObject *other)
+{
+    Py_ssize_t pos1 = 0,
+               pos2 = 0;
+
+    Py_hash_t h1 = 0,
+              h2 = 0;
+
+    PyObject *identity1 = NULL,
+             *identity2 = NULL,
+             *value1    = NULL,
+             *value2    = NULL;
+
+    if (self == other) {
+        Py_RETURN_TRUE;
+    }
+
+    if (pair_list_len(self->impl) != pair_list_len(other->impl)) {
+        Py_RETURN_FALSE;
+    }
+
+    while (_pair_list_next(self->impl, &pos1, &identity1, NULL, &value1, &h1) &&
+           _pair_list_next(other->impl, &pos2, &identity2, NULL, &value2, &h2))
+    {
+        if ((h1 != h2) ||
+            PyObject_RichCompare(identity1, identity2, Py_NE) ||
+            PyObject_RichCompare(value1, value2, Py_NE))
+        {
+            Py_RETURN_FALSE;
+        }
+    }
+
+    Py_RETURN_TRUE;
+}
+
+static INLINE int
+_multidict_update_items(_MultiDictObject *self, _MultiDictObject *impl)
+{
+    return pair_list_update((PyObject*)self->impl, (PyObject*)impl);
+}
+
+static int
+_multidict_append_items(_MultiDictObject *self, _MultiDictObject *impl)
+{
+    PyObject *key   = NULL,
+             *value = NULL;
+
+    Py_ssize_t pos = 0;
+
+    while (_pair_list_next((PyObject*)impl, &pos, NULL, &key, &value, NULL)) {
+        if (pair_list_add(self->impl, key, value) < 0) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int
+_multidict_append_items_seq(_MultiDictObject *self, PyObject *arg,
+                            const char *name)
+{
+    PyObject *key   = NULL,
+             *value = NULL,
+             *item  = NULL,
+             *iter  = PyObject_GetIter(arg);
+
+    if (iter == NULL) {
+        return -1;
+    }
+
+    while ((item = PyIter_Next(iter)) != NULL) {
+        if (PyObject_Length(item) != 2) {
+            PyErr_Format(
+                PyExc_TypeError,
+                "%s takes either dict or list of (key, value) tuples",
+                name,
+                NULL
+            );
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+
+        key   = PyTuple_GET_ITEM(item, 0);
+        value = PyTuple_GET_ITEM(item, 0);
+        pair_list_add(self->impl, key, value);
+
+        Py_DECREF(item);
+    }
+
```

Comment:
`PyIter_Next()` can return `NULL` **and** set an exception if an error occured.
Should return non-null if `PyErr_Occured()` returns non-zero

Reply:
Please apply the check to all `PyIter_Next()` calls in the file

Reply:
Done